### PR TITLE
Make prometheus modular[WIP]

### PR DESCRIPTION
**Pull request by hskang9**

### CODE REVIEW

File: core/prometheus/src/metrics.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+pub use crate::*;
+
+lazy_static! {
+    /// Global metrics to put in prometheus
+    pub static ref HEIGHT: Result<IntGauge> = try_create_int_gauge(
+        "libp2p_peer_connected_peers_total",
+        "Count of libp2p peers currently connected"
+    );
+}
```
Comment (DemiMarie-parity) [inline core/prometheus/src/metrics.rs]: nit: missing newline at EOF

File: core/prometheus/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    std::thread::spawn(move || {
+        rt.spawn(server);
+        rt.shutdown_on_idle().wait().unwrap();
+    });
+}
+
+
+//impl trait<Number> for dyn Header {
+//    type Number = u64;
+//}
```
Comment (DemiMarie-parity) [inline core/prometheus/src/lib.rs]: nit: missing newline at EOF

File: core/prometheus/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                    let metric_families = prometheus::gather();
+                    let mut buffer = vec![];
+                    let encoder = TextEncoder::new();
+                    encoder.encode(&metric_families, &mut buffer).unwrap();
+                    HTTP_BODY_GAUGE.set(buffer.len() as f64);
+                    Response::builder()
+                        .status(StatusCode::OK)
+                        .header("Content-Type", encoder.format_type())
+                        .body(Body::from(buffer))
+                        .expect("Error constructing response")
```
Comment (DemiMarie-parity) [inline core/prometheus/src/lib.rs]: Insufficient justification for panic

File: core/prometheus/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                } else {
+                    Response::builder()
+                        .status(StatusCode::NOT_FOUND)
+                        .body(Body::from("Not found."))
+                        .expect("Error constructing response")
```
Comment (DemiMarie-parity) [inline core/prometheus/src/lib.rs]: Insufficient justification for panic

**Comment by hskang9** [comment]
Thank you for the review and I will make sure I add that. However, the working branch has moved from here to `nodebreaker-carlsagan:prometheus_v0.1` due to substrate v2.0 upgrade. I will add new PRs to this.

File: core/prometheus/src/lib.rs (hunk 3 above)
Comment (hskang9) [reply core/prometheus/src/lib.rs]: Done at 
https://github.com/nodebreaker-carlsagan/substrate/commit/c48bad6a823bfe62043fb0313ddbdc46815ee9a4

File: core/prometheus/src/lib.rs (hunk 4 above)
Comment (hskang9) [reply core/prometheus/src/lib.rs]: Done at
https://github.com/nodebreaker-carlsagan/substrate/commit/c48bad6a823bfe62043fb0313ddbdc46815ee9a4

File: core/prometheus/src/lib.rs (hunk 2 above)
Comment (hskang9) [reply core/prometheus/src/lib.rs]: Done at 
https://github.com/nodebreaker-carlsagan/substrate/commit/c48bad6a823bfe62043fb0313ddbdc46815ee9a4

File: core/prometheus/src/metrics.rs (hunk 1 above)
Comment (hskang9) [reply core/prometheus/src/metrics.rs]: Done at 
https://github.com/nodebreaker-carlsagan/substrate/commit/c48bad6a823bfe62043fb0313ddbdc46815ee9a4
Please review.

File: core/prometheus/src/lib.rs (hunk 4 above)
Comment (DemiMarie-parity) [reply core/prometheus/src/lib.rs]: Looks good @hskang9.  Please push to the current branch so it shows up in the PR UI.

File: core/prometheus/src/lib.rs (hunk 3 above)
Comment (DemiMarie-parity) [reply core/prometheus/src/lib.rs]: Looks good @hskang9.  Please push to the current branch so it shows up in the PR UI.

File: core/prometheus/src/lib.rs (hunk 2 above)
Comment (DemiMarie-parity) [reply core/prometheus/src/lib.rs]: Looks good @hskang9.  Please push to the current branch so it shows up in the PR UI.

Comment (hskang9) [reply core/prometheus/src/lib.rs]: Thanks. Once I finish runtime implementation on prometheus, I will add changes here.

**Comment by bkchr** [comment]
@hskang9 any update?

**Comment by hskang9** [comment]
Now Prometheus supports macros for data types to return `Gauge`, `Histogram`, and `Counter` as telemetry does it with `telemetry!` [here](https://github.com/nodebreaker0-0/substrate/blob/b7fd6b334712f3a8fc5bbb7a63a70c54fbaec356/client/service/src/builder.rs#L958). Unifying them with polymorphism is on-going. I am trying to make prometheus macro work in runtime with offchain worker callback, collecting event values and send it. Package info is [here](https://github.com/nodebreaker0-0/substrate/tree/prometheus_v0.3/client/prometheus). After working on runtime, I will integrate it to this current branch.

**Comment by bkchr** [comment]
Superseded by: https://github.com/paritytech/substrate/pull/4511

**Pull request closed without merging**